core/server/worker/src/main/java/alluxio/worker/page/PagedBlockReader.java

```diff
@@ -109,48 +116,60 @@ public ByteBuffer read(long offset, long length) throws IOException {
           throw new IOException(String.format("Block %d does not have UFS read options, "
               + "therefore cannot be read from UFS", mBlockMeta.getBlockId()));
         }
-        byte[] page = readPageFromUFS(mUfsBlockOptions.get(), pos);
-        if (page.length > 0) {
-          System.arraycopy(page, currentPageOffset, buf, (int) bytesRead, bytesLeftInPage);
-          bytesRead += bytesLeftInPage;
-          MetricsSystem.meter(MetricKey.CLIENT_CACHE_BYTES_REQUESTED_EXTERNAL.getName())
-              .mark(bytesLeftInPage);
-          mReadFromUfs = true;
-          mCacheManager.put(pageId, page);
+        long pageStart = pos - (pos % mPageSize);
+        int pageSize = (int) Math.min(mPageSize, mBlockMeta.getBlockSize() - pageStart);
+        ByteBuffer ufsBuf = NioDirectBufferPool.acquire(pageSize);
+        try {
+          int bytesReadFromUFS = readPageFromUFS(pageStart, pageSize, ufsBuf);
+          if (ufsBuf.remaining() > 0) {
+            ufsBuf.mark();
+            ufsBuf.position(currentPageOffset);
+            ufsBuf.limit(currentPageOffset + bytesLeftInPage);
+            buf.put(ufsBuf);
+            bytesRead += bytesLeftInPage;
+            MetricsSystem.meter(MetricKey.CLIENT_CACHE_BYTES_REQUESTED_EXTERNAL.getName())
+                .mark(bytesLeftInPage);
+            mReadFromUfs = true;
+            ufsBuf.reset();
+            ufsBuf.limit(bytesReadFromUFS);
+            mCacheManager.put(pageId, ufsBuf);
+          }
+        } finally {
+          NioDirectBufferPool.release(ufsBuf);
         }
       }
     }
-    return ByteBuffer.wrap(buf);
+    buf.flip();
+    return buf;
   }
 
-  private byte[] readPageFromUFS(UfsBlockReadOptions options, long pos) throws IOException {
-    long pageStart = pos - (pos % mPageSize);
-    InputStream ufsInputStream = seekUfsInputStream(options, options.getOffsetInFile() + pageStart);
-    int pageSize = (int) Math.min(mPageSize, mBlockMeta.getBlockSize() - pageStart);
-    byte[] page = new byte[pageSize];
+  private int readPageFromUFS(long pageStart, int pageSize, ByteBuffer target) throws IOException {
+    InputStream ufsInputStream =
+        seekUfsInputStream(mUfsBlockOptions.get().getOffsetInFile() + pageStart);
     int totalBytesRead = 0;
     try {
       while (totalBytesRead < pageSize) {
-        int bytesRead = ufsInputStream.read(page, totalBytesRead, pageSize - totalBytesRead);
+        int bytesRead = Channels.newChannel(ufsInputStream).read(target);
```

Comment:
just a technical question. What would happen if there's too many channels opening for this ufsInputStream? Would those get GC eventually if ufsInputStream is cached?

Reply:
As we just discussed offline, yes, it will be recycled by GC